### DISCUSSION TITLE
HOW TO CONNECT TO PROVIDER IN WEB3MODAL V2.0

### DISCUSSION THREAD

**Discussion by EECvision** [WalletConnect/walletconnect-monorepo#3443] [Developer Support] [answered]
in V1 I get provider by doing
```
const web3Provider = new WalletConnectProvider({
  infuraId: 'INFURA_ID'
})

provider = new ethers.providers.Web3Provider(web3Provider);
```
Now that I'm using Web3Modal v2.0 
```
import { w3mConnectors, w3mProvider } from '@web3modal/ethereum'
import { createConfig } from 'wagmi'

// ...
const { publicClient } = configureChains(chains, [w3mProvider({ projectId })])

const config = createConfig({
  autoConnect: true,
  connectors: [
    w3mConnectors({
      projectId,
      chains
    })
  ],
  publicClient
})

const ethereumClient = new EthereumClient(config, chains)
// ...
```

How do I achieve the same result as v1? How do I set my provider to be able to sign transactions with it.

**Comment by glitch-txs** [comment, marked as answer]
Web3Modal v2 works with [Wagmi](https://wagmi.sh/) you can learn more of its implementation on their [docs](https://wagmi.sh/react/getting-started).

**Reply by Encodtech** [reply]
@glitch-txs Is it possible to get connected accounts on Web3Modal using ==> accounts = await web3.eth.requestAccounts() ?
I was hoping I could define the Provider so I could do actions like ==>  accounts = await provider.request({method: "eth_requestAccounts"});

The Wagmi and Viem actions to get connected accounts works on perfectly Metamask but not on Trustwallet.

**Reply by glitch-txs** [reply]
What's the issue with Trust Wallet?

You can try replacing ethers with web3js [here](https://wagmi.sh/core/ethers-adapters)

**Reply by Encodtech** [reply]
@glitch-txs On Trustwallet, it throws an error on the console, "account and/or method has not been authorized by the user". Maybe to say that there's no connected account for that action.

**Reply by glitch-txs** [reply]
That's happens when you're not using the actual WalletConnect provider. Could you share a minimal example on a repo or StackBlitz?

**Reply by Encodtech** [reply]
@glitch-txs is there a CDN method to call the Walletconnect provider. My project is in vanilla JavaScript. Any examples would be greatly appreciated.

**Reply by glitch-txs** [reply]
atm only with wagmi actions https://wagmi.sh/core/actions/writeContract

You can also try using their [adapter](https://wagmi.sh/core/ethers-adapters)

**Reply by Encodtech** [reply]
@glitch-txs I just realized Trustwallet throws that error because somehow Wagmi expects my sendTransaction action to happen in connected Metamask even when Web3Modal is connected to Trustwallet. How can I fix this please?

**Reply by glitch-txs** [reply]
please open a new discussion also could you share there a snippet of your code in a public repo/stackBlitz?

**Reply by DarioVincenzoDeSimone** [reply]
Maybe, i found a solution using wagmi method:


```
import { getAccount } from "@wagmi/core";
let provider = getAccount().connector.options.getProvider();
```

...but it works only if you select wallet with eg. metamask chrome extension.
If you sign with QRcode getProvider() method doesn't exist and i really don't know why and how to retrive provider.

**Reply by glitch-txs** [reply]
maybe try using `await`

**Reply by DarioVincenzoDeSimone** [reply]
I don't think `await` is necessary, `getAccount()` is not a promise.
The object `getAccount().connector.options` contain method getProvider() only if you select wallet from extension.

I found a (not so clean) solution:

```
try {
provider = account.connector.options?.getProvider();
} catch (error) {
provider = null;
}
```
if  `provider == null` it means that you select wallet using QRcode and you can retrive provider in this condition using `await EthereumProvider.init(providerOption)`.

I hope someone else find a better solution for it.

**Reply by glitch-txs** [reply]
getProvider can be is async

**Reply by oyin25** [reply]
Better way to get the provider, 
Here is what i aims to achieve,
I am using sendTransaction to send a transaction but the connected wallet window will not open itself to trigger the transactions on mobile wallet, though everything work fine on Desktop PC.
